crates/matrix-sdk-crypto/src/machine/mod.rs

```diff
@@ -1336,7 +1359,54 @@ impl OlmMachine {
             e => self.handle_to_device_event(changes, &e).await,
         }
 
-        raw_event
+        Some(raw_event)
+    }
+
+    async fn event_is_from_dehydrated_device(
+        &self,
+        decrypted: &OlmDecryptionInfo,
+        sender_user_id: &UserId,
+    ) -> bool {
+        // Does the to-device message include device info?
+        if let Some(device_keys) = decrypted.result.event.sender_device_keys() {
+            // There is no need to check whether the device keys are signed correctly - any
+            // to-device message that claims to be from a dehydrated device is weird, so we
+            // will drop it.
+
+            // Does the included device info say the device is dehydrated?
+            if device_keys.dehydrated.unwrap_or(false) {
+                return true;
+            }
+            // If not, fall through and check our existing list of devices
+            // below, just in case the sender is sending us
+            // incorrect information embedded in the to-device message, but we
+            // know better.
+        }
+
+        // Do we already know about this device?
+        match self
+            .store()
+            .get_device_from_curve_key(sender_user_id, decrypted.result.sender_key)
+            .await
+        {
+            Ok(device) => {
+                // Yes - is it dehydrated?
+                if let Some(device) = device {
+                    device.is_dehydrated()
+                } else {
+                    false
+                }
```

Comment:
light suggestion, no strong feelings here

```suggestion
                device.is_some_and(Device::is_dehydrated)
```

Reply:
Changed in e7eaa3e539 (exact suggestion didn't because it would have need an `as_ref` (I think))